contracts/SpokePool.sol

```diff
@@ -1373,24 +1397,41 @@ abstract contract SpokePool is
         uint32 leafId,
         bytes32 l2TokenAddress,
         bytes32[] memory refundAddresses
-    ) internal {
+    ) internal returns (bool deferredRefunds) {
         if (refundAddresses.length != refundAmounts.length) revert InvalidMerkleLeaf();
 
-        address l2TokenAddressParsed = l2TokenAddress.toAddress();
+        uint256 spokeStartBalance = IERC20Upgradeable(l2TokenAddress.toAddress()).balanceOf(address(this));
+        uint256 totalRefundedAmount = 0;
 
         // Send each relayer refund address the associated refundAmount for the L2 token address.
         // Note: Even if the L2 token is not enabled on this spoke pool, we should still refund relayers.
-        uint256 length = refundAmounts.length;
-        for (uint256 i = 0; i < length; ++i) {
-            uint256 amount = refundAmounts[i];
-            if (amount > 0)
-                IERC20Upgradeable(l2TokenAddressParsed).safeTransfer(refundAddresses[i].toAddress(), amount);
+        for (uint256 i = 0; i < refundAmounts.length; ++i) {
+            totalRefundedAmount += refundAmounts[i];
+
+            if (totalRefundedAmount > spokeStartBalance) revert InsufficientSpokePoolBalanceToExecuteLeaf();
+            if (refundAmounts[i] > 0) {
+                bool success;
+                try
+                    IERC20Upgradeable(l2TokenAddress.toAddress()).transfer(
+                        refundAddresses[i].toAddress(),
+                        refundAmounts[i]
+                    )
+                returns (bool result) {
```

Comment:
This might fail on successful transfers for tokens that do not return bool on success (e.g. USDT). So instead of try/catch pattern we might need to reimplement OZ `safeTransfer` without bubbling up the reverts.

Reply:
yup, I agree. we should just catch on reverts. what do you think of this implementation?

else, you're right: we can implement [this](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/c343ee3768cc61a76a14d6f8402b1b19be1497fc/contracts/token/ERC20/utils/SafeERC20.sol#L35) kind of structure internally

Reply:
the safe erc20 assumes `non-reverting calls are assumed to be successful`, which is what the new implementation now assumes.

Reply:
`IERC20Upgradeable` expects bool to be returned from the transfer, so USDT would always hit the catch block. Since try/catch works only on external calls or contract creation, the only viable option is to replicate the logic of OZ `_callOptionalReturn` and replacing all reverts with returning bool that we can track in our logic.